documentation/library-reference/source/io/streams.rst

```diff
@@ -464,72 +344,160 @@ streams while the stream has it open. The possible values are:
+
+A common way to ensure that :gf:`close` is called is to use the ``cleanup``
+clause of the :drm:`block` macro. For example:
+
+.. code-block:: dylan
+
+   let stream = open-my-stream(...);
+   block ()
+     ...read or write to stream...
+   cleanup
+     close(stream);
+   end
+
+To make this easier, and so you're less likely to forget to call :gf:`close`,
+most libraries that provide stream classes provide a macro. For example,
+:macro:`with-open-file` for file streams, ensures that :gf:`close` is called:
+
+.. code-block:: dylan
+
+   with-open-file (stream = "/my/file")
+     ...read or write to stream...
+     // close(stream) is called automatically
+   end
+
 
 Locking streams
 ^^^^^^^^^^^^^^^
 
-In an application where more than one control thread may access a common
-stream, it is important to match the granularity of locking to the
-transaction model of the application. Ideally, an application should
-lock a stream which is potentially accessed by multiple threads, only
-once per transaction. Repeated and unnecessary locking and unlocking can
-seriously degrade the performance of the Streams module. Thus an
-application which wishes to write a complex message to a stream that
-needs to be thread safe should lock the stream, write the message and
-then unlock the stream after the entire message is written. Locking and
-unlocking the stream for each character in the message would be a poor
-match of locking to transaction model. The time required for the lock
-manipulation would dominate the time required for the stream
-transactions. Unfortunately this means that there is no way for the
-Streams module to choose a default locking scheme without the likelihood
-of seriously degrading streams performance for all applications whose
-transaction models are different from the model implied by the chosen
-default locking scheme. Instead, the Streams module provides the user
-with a single, per instance slot, *stream-lock:*, which is inherited by
-all subclasses of :class:`<stream>`. You should use the generic functions
-:gf:`lock-stream` and :gf:`unlock-stream` or the macro
-:macro:`with-stream-locked`, together with other appropriate functions
-and macros from the Threads library, to implement a locking strategy
-appropriate to your application and its stream transaction model. The
-functions in the Streams module are not of themselves thread safe,
-and make no guarantees about the atomicity of read and write operations.
+In an application where more than one thread may access a shared stream, it is
+important to match the granularity of locking to the transaction model of the
+application. Ideally, an application should lock a stream only once per
+transaction, to minimize the performance penalty. Since the streams module
+cannot know the transaction model of your code, it does not provide any
+default locking scheme.
+
+Instead, the streams module provides the user with a single, per instance slot,
```

Comment:
Perhaps it's not obvious from this paragraph that you need to supply e.g. `stream-lock: make(<recursive-lock>)` when instantiating the stream

Reply:
Done.